
.. DO NOT EDIT.
.. THIS FILE WAS AUTOMATICALLY GENERATED BY SPHINX-GALLERY.
.. TO MAKE CHANGES, EDIT THE SOURCE PYTHON FILE:
.. "examples_gallery/plot_position_changes.py"
.. LINE NUMBERS ARE GIVEN BELOW.

.. only:: html

    .. note::
        :class: sphx-glr-download-link-note

        :ref:`Go to the end <sphx_glr_download_examples_gallery_plot_position_changes.py>`
        to download the full example code

.. rst-class:: sphx-glr-example-title

.. _sphx_glr_examples_gallery_plot_position_changes.py:


Position changes during a race
==============================

Plot the position of each driver at the end of each lap.

.. GENERATED FROM PYTHON SOURCE LINES 7-14

.. code-block:: default


    import fastf1.plotting
    import matplotlib.pyplot as plt


    fastf1.plotting.setup_mpl(misc_mpl_mods=False)








.. GENERATED FROM PYTHON SOURCE LINES 15-16

Load the session and create the plot

.. GENERATED FROM PYTHON SOURCE LINES 16-21

.. code-block:: default

    session = fastf1.get_session(2023, 1, 'R')
    session.load(telemetry=False, weather=False)

    fig, ax = plt.subplots(figsize=(8.0, 4.9))





.. rst-class:: sphx-glr-script-out

 .. code-block:: none

    Request for URL https://ergast.com/api/f1/2023/1/laps/1.json failed; using cached response
    Traceback (most recent call last):
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/urllib3/connectionpool.py", line 790, in urlopen
        response = self._make_request(
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/urllib3/connectionpool.py", line 536, in _make_request
        response = conn.getresponse()
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/urllib3/connection.py", line 461, in getresponse
        httplib_response = super().getresponse()
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/http/client.py", line 1377, in getresponse
        response.begin()
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/http/client.py", line 320, in begin
        version, status, reason = self._read_status()
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/http/client.py", line 289, in _read_status
        raise RemoteDisconnected("Remote end closed connection without"
    http.client.RemoteDisconnected: Remote end closed connection without response

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/requests/adapters.py", line 486, in send
        resp = conn.urlopen(
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/urllib3/connectionpool.py", line 844, in urlopen
        retries = retries.increment(
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/urllib3/util/retry.py", line 470, in increment
        raise reraise(type(error), error, _stacktrace)
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/urllib3/util/util.py", line 38, in reraise
        raise value.with_traceback(tb)
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/urllib3/connectionpool.py", line 790, in urlopen
        response = self._make_request(
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/urllib3/connectionpool.py", line 536, in _make_request
        response = conn.getresponse()
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/urllib3/connection.py", line 461, in getresponse
        httplib_response = super().getresponse()
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/http/client.py", line 1377, in getresponse
        response.begin()
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/http/client.py", line 320, in begin
        version, status, reason = self._read_status()
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/http/client.py", line 289, in _read_status
        raise RemoteDisconnected("Remote end closed connection without"
    urllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/requests_cache/session.py", line 255, in _resend
        response = self._send_and_cache(request, actions, cached_response, **kwargs)
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/requests_cache/session.py", line 229, in _send_and_cache
        response = super().send(request, **kwargs)
      File "/home/runner/work/Fast-F1/Fast-F1/fastf1/req.py", line 124, in send
        return super().send(request, **kwargs)
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/requests/sessions.py", line 703, in send
        r = adapter.send(request, **kwargs)
      File "/opt/hostedtoolcache/Python/3.9.18/x64/lib/python3.9/site-packages/requests/adapters.py", line 501, in send
        raise ConnectionError(err, request=request)
    requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))




.. GENERATED FROM PYTHON SOURCE LINES 23-26

For each driver, get their three letter abbreviation (e.g. 'HAM') by simply
using the value of the first lap, get their color and then plot their
position over the number of laps.

.. GENERATED FROM PYTHON SOURCE LINES 26-35

.. code-block:: default

    for drv in session.drivers:
        drv_laps = session.laps.pick_driver(drv)

        abb = drv_laps['Driver'].iloc[0]
        color = fastf1.plotting.driver_color(abb)

        ax.plot(drv_laps['LapNumber'], drv_laps['Position'],
                label=abb, color=color)








.. GENERATED FROM PYTHON SOURCE LINES 37-39

Finalize the plot by setting y-limits that invert the y-axis so that position
one is at the top, set custom tick positions and axis labels.

.. GENERATED FROM PYTHON SOURCE LINES 39-44

.. code-block:: default

    ax.set_ylim([20.5, 0.5])
    ax.set_yticks([1, 5, 10, 15, 20])
    ax.set_xlabel('Lap')
    ax.set_ylabel('Position')





.. rst-class:: sphx-glr-script-out

 .. code-block:: none


    Text(0, 0.5, 'Position')



.. GENERATED FROM PYTHON SOURCE LINES 46-47

Because this plot is very crowed, add the legend outside the plot area.

.. GENERATED FROM PYTHON SOURCE LINES 47-51

.. code-block:: default

    ax.legend(bbox_to_anchor=(1.0, 1.02))
    plt.tight_layout()

    plt.show()



.. image-sg:: /examples_gallery/images/sphx_glr_plot_position_changes_001.png
   :alt: plot position changes
   :srcset: /examples_gallery/images/sphx_glr_plot_position_changes_001.png
   :class: sphx-glr-single-img






.. rst-class:: sphx-glr-timing

   **Total running time of the script:** (0 minutes 7.884 seconds)


.. _sphx_glr_download_examples_gallery_plot_position_changes.py:

.. only:: html

  .. container:: sphx-glr-footer sphx-glr-footer-example




    .. container:: sphx-glr-download sphx-glr-download-python

      :download:`Download Python source code: plot_position_changes.py <plot_position_changes.py>`

    .. container:: sphx-glr-download sphx-glr-download-jupyter

      :download:`Download Jupyter notebook: plot_position_changes.ipynb <plot_position_changes.ipynb>`


.. only:: html

 .. rst-class:: sphx-glr-signature

    `Gallery generated by Sphinx-Gallery <https://sphinx-gallery.github.io>`_
